File: 714-best-time-to-buy-and-sell-stock-with-transaction-fee/714-best-time-to-buy-and-sell-stock-with-transaction-fee.py
class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        minVal = prices[0]
        returnVal = 0
        for i in range(1,len(prices)):
            if prices[i] < minVal:
                minVal = prices[i]
            else:
                if prices[i] - minVal <= fee:
                    continue
                else:
                    returnVal += prices[i] - minVal - fee
                    minVal = prices[i] - fee
        return returnVal
        

File: 714-best-time-to-buy-and-sell-stock-with-transaction-fee/714-best-time-to-buy-and-sell-stock-with-transaction-fee.java
class Solution {
    public int maxProfit(int[] prices, int fee) {
        int buy = - prices[0] , sell = 0 , tempBuy , tempSell;
        for (int i = 1 ; i < prices.length ; i ++){
            tempBuy = buy ; 
            tempSell = sell ; 
            buy = Math.max(tempBuy,tempSell - prices[i]) ; 
            sell = Math.max(tempSell , tempBuy + prices[i] - fee) ; 
        }
        return sell ; 
    }
}